### PR TITLE
ENH: implement right multiplication of linear operators by arrays

### PR DESCRIPTION
**Pull request by akhmerov**
#### Reference issue

N/A

#### What does this implement/fix?

Allows to multiply a linear operator by a numpy array from the left if the linear operator supports this.

### CODE REVIEW

File: scipy/sparse/linalg/tests/test_interface.py [hunk 1]
```diff
@@ -448,3 +448,8 @@ def test_transpose_noconjugate():
 
     assert_equal(B.dot(v), Y.dot(v))
     assert_equal(B.T.dot(v), Y.T.dot(v))
+
```
Comment (akhmerov) [inline scipy/sparse/linalg/tests/test_interface.py:451]: Linter complains about a single and not double blank line in tests, however I copied the formatting from surrounding code. Should I change this to two lines?

Comment (tupui) [reply scipy/sparse/linalg/tests/test_interface.py:451]: Yes 2 lines are required between standalone functions. The linter is just looking at current changes. This is why it cannot spot old mistakes 😅

File: scipy/sparse/linalg/interface.py [hunk 2]
```diff
@@ -437,9 +438,18 @@ def __rmatmul__(self, other):
...
             return NotImplemented
 
+        x = np.asarray(x).conj()
+
+        if x.ndim == 1 or x.ndim == 2 and x.shape[0] == 1:
+            return self.rmatvec(x).conj()
+        elif x.ndim == 2:
+            return self.rmatmat(x).conj()
+        else:
+            raise ValueError(f'expected 1-d or 2-d array or matrix, got {x}')
```
Comment (tupui) [inline scipy/sparse/linalg/interface.py:451]: This also need a test with `with pytest.raises(ValueError, match="expected 1-d or 2-d array")`. This way you ensure you are catching the right error.

File: scipy/sparse/linalg/tests/test_interface.py (hunk 1 above)
Comment (akhmerov) [reply scipy/sparse/linalg/tests/test_interface.py:451]: I understand, but what is the preferred way to go:

- apply the new style and have inconsistent formatting
- leave 1 line
- fix this part of formatting in the whole module in a separate commit

Comment (tupui) [reply scipy/sparse/linalg/tests/test_interface.py:451]: For sure any new code should pass the linter/tests. Usually we do not want people to reformat what they don't touch as to not break the git history and blame. If it's just a few things fine it's ok, but if the diff is too large, please don't change it.

Comment (akhmerov) [reply scipy/sparse/linalg/tests/test_interface.py:451]: For now I went with the first option to make tests pass, however happy to revert to either of the other alternatives.

**Comment by tupui** [comment]
@jakevdp I think this is your area of expertise, if you have some time.

**Comment by tupuinui** [comment]
FYI @akhmerov, usually we prefer not to force push until the end of the review as it tends to break links in GitHub (clicking View changes does not work then). Before merging we always make sure the history is not too cloudy and in worst case we squash merge the commits. No worries though here it's fine 😉

**Comment by tupui** [comment]
FYI @akhmerov, usually we prefer not to force push until the end of the review as it tends to break links in GitHub (clicking View changes does not work then). Before merging we always make sure the history is not too cloudy and in worst case we squash merge the commits. No worries though here it's fine 😉

**Comment by akhmerov** [comment]
> usually we prefer not to force push until the end of the review as it tends to break links in GitHub

Thanks for the heads up, I'll take this into account next time.

**Comment by akhmerov** [comment]
Should be good to go, pardon for the noise with the bug/typo fixes.

**Comment by akhmerov** [comment]
This is superseded by #18061 and obsolete. Closing.

**Pull request closed without merging**